File: apps/macos/Sources/TrimletCore/EditList.swift
import Foundation

public struct EditSegment: Identifiable, Codable, Hashable, Sendable {
    public let id: UUID
    public var inPoint: MediaTimestamp
    public var outPoint: MediaTimestamp
    public var clipNumber: Int?

    public init(
        id: UUID = UUID(),
        inPoint: MediaTimestamp,
        outPoint: MediaTimestamp,
        clipNumber: Int? = nil
    ) {
        self.id = id
        self.inPoint = inPoint
        self.outPoint = outPoint
        self.clipNumber = clipNumber
    }

    public init(id: UUID = UUID(), range: TrimRange, clipNumber: Int? = nil) throws {
        guard let start = range.inPoint,
              let end = range.outPoint,
              end > start else {
            throw EditListError.invalidRange
        }
        self.init(
            id: id,
            inPoint: MediaTimestamp(seconds: start),
            outPoint: MediaTimestamp(seconds: end),
            clipNumber: clipNumber
        )
    }

    public var isValid: Bool { outPoint > inPoint }
    public var durationSeconds: Double { max(0, outPoint.seconds - inPoint.seconds) }
    public var trimRange: TrimRange {
        TrimRange(inPoint: inPoint.seconds, outPoint: outPoint.seconds)
    }

    public func overlaps(_ other: EditSegment) -> Bool {
        inPoint < other.outPoint && other.inPoint < outPoint
    }
}

public struct EditList: Codable, Equatable, Sendable {
    public private(set) var segments: [EditSegment]

    public init() {
        segments = []
    }

    public init(segments: [EditSegment]) throws {
        self.segments = []
        for segment in segments {
            try append(segment)
        }
    }

    public var isEmpty: Bool { segments.isEmpty }
    public var totalDurationSeconds: Double {
        segments.reduce(0) { $0 + $1.durationSeconds }
    }

    public func segment(id: UUID?) -> EditSegment? {
        guard let id else { return nil }
        return segments.first { $0.id == id }
    }

    public mutating func append(_ segment: EditSegment, sourceDuration: MediaTimestamp? = nil) throws {
        try validate(segment, replacing: nil, sourceDuration: sourceDuration)
        segments.append(segment)
    }

    public mutating func update(_ segment: EditSegment, sourceDuration: MediaTimestamp? = nil) throws {
        guard let index = segments.firstIndex(where: { $0.id == segment.id }) else {
            throw EditListError.segmentNotFound
        }
        try validate(segment, replacing: segment.id, sourceDuration: sourceDuration)
        segments[index] = segment
    }

    @discardableResult
    public mutating func remove(id: UUID) throws -> EditSegment {
        guard let index = segments.firstIndex(where: { $0.id == id }) else {
            throw EditListError.segmentNotFound
        }
        return segments.remove(at: index)
    }

    public mutating func move(id: UUID, by offset: Int) throws {
        guard let from = segments.firstIndex(where: { $0.id == id }) else {
            throw EditListError.segmentNotFound
        }
        let destination = min(max(0, from + offset), segments.count - 1)
        guard destination != from else { return }
        let segment = segments.remove(at: from)
        segments.insert(segment, at: destination)
    }

    public mutating func move(id: UUID, to destinationIndex: Int) throws {
        guard let from = segments.firstIndex(where: { $0.id == id }) else {
            throw EditListError.segmentNotFound
        }
        let segment = segments.remove(at: from)
        let destination = min(max(0, destinationIndex), segments.count)
        segments.insert(segment, at: destination)
    }

    public func validate(sourceDuration: MediaTimestamp? = nil) throws {
        for segment in segments {
            try validate(segment, replacing: segment.id, sourceDuration: sourceDuration)
        }
    }

    private func validate(
        _ segment: EditSegment,
        replacing id: UUID?,
        sourceDuration: MediaTimestamp?
    ) throws {
        guard segment.isValid else { throw EditListError.invalidRange }
        if let sourceDuration, segment.outPoint > sourceDuration {
            throw EditListError.outsideSource
        }
        if segments.contains(where: { existing in
            existing.id != id && existing.overlaps(segment)
        }) {
            throw EditListError.overlap
        }
    }
}

public enum EditListError: LocalizedError, Equatable, Sendable {
    case invalidRange
    case outsideSource
    case overlap
    case segmentNotFound

    public var errorDescription: String? {
        switch self {
        case .invalidRange:
            "IN点をOUT点より前に設定してください。"
        case .outsideSource:
            "クリップが元動画の範囲外です。"
        case .overlap:
            "サブクリップが既存のクリップと重なっています。"
        case .segmentNotFound:
            "クリップが見つかりません。"
        }
    }
}

File: apps/windows/src/Trimlet.Media/EditList.cs
namespace Trimlet.Media;

public sealed record EditSegment(Guid Id, string Name, TrimRange Range)
{
    public double DurationSeconds => Range.DurationSeconds;

    public bool Overlaps(EditSegment other) =>
        Range.In < other.Range.Out && other.Range.In < Range.Out;

    public EditSegment WithRange(TrimRange range) => this with { Range = range };

    public EditSegment WithName(string name) => this with
    {
        Name = string.IsNullOrWhiteSpace(name) ? Name : name.Trim(),
    };
}

public sealed class EditList : IEquatable<EditList>
{
    private readonly EditSegment[] _segments;

    public EditList() : this([], validate: false)
    {
    }

    public EditList(IEnumerable<EditSegment> segments) : this(segments.ToArray(), validate: true)
    {
    }

    private EditList(EditSegment[] segments, bool validate)
    {
        _segments = [.. segments];
        if (validate)
        {
            Validate();
        }
    }

    public IReadOnlyList<EditSegment> Segments => _segments;
    public bool IsEmpty => _segments.Length == 0;
    public double TotalDurationSeconds => _segments.Sum(segment => segment.DurationSeconds);

    public EditSegment? Segment(Guid id) => _segments.FirstOrDefault(segment => segment.Id == id);

    public EditList Add(EditSegment segment, MediaTimestamp? sourceDuration = null)
    {
        ValidateSegment(segment, null, sourceDuration);
        return new EditList([.. _segments, segment], validate: false);
    }

    public EditList Update(EditSegment segment, MediaTimestamp? sourceDuration = null)
    {
        var index = Array.FindIndex(_segments, existing => existing.Id == segment.Id);
        if (index < 0)
        {
            throw new KeyNotFoundException("The clip was not found.");
        }

        ValidateSegment(segment, segment.Id, sourceDuration);
        var updated = _segments.ToArray();
        updated[index] = segment;
        return new EditList(updated, validate: false);
    }

    public EditList Remove(Guid id)
    {
        if (!_segments.Any(segment => segment.Id == id))
        {
            throw new KeyNotFoundException("The clip was not found.");
        }

        return new EditList(_segments.Where(segment => segment.Id != id).ToArray(), validate: false);
    }

    public EditList Move(Guid id, int destinationIndex)
    {
        var sourceIndex = Array.FindIndex(_segments, segment => segment.Id == id);
        if (sourceIndex < 0)
        {
            throw new KeyNotFoundException("The clip was not found.");
        }

        var reordered = _segments.ToList();
        var segment = reordered[sourceIndex];
        reordered.RemoveAt(sourceIndex);
        reordered.Insert(Math.Clamp(destinationIndex, 0, reordered.Count), segment);
        return new EditList([.. reordered], validate: false);
    }

    public void Validate(MediaTimestamp? sourceDuration = null)
    {
        var identifiers = new HashSet<Guid>();
        foreach (var segment in _segments)
        {
            if (!identifiers.Add(segment.Id))
            {
                throw new InvalidDataException("Clip identifiers must be unique.");
            }

            ValidateSegment(segment, segment.Id, sourceDuration);
        }
    }

    private void ValidateSegment(EditSegment segment, Guid? replacing, MediaTimestamp? sourceDuration)
    {
        if (segment.Id == Guid.Empty)
        {
            throw new InvalidDataException("A clip must have a stable identifier.");
        }

        if (string.IsNullOrWhiteSpace(segment.Name))
        {
            throw new InvalidDataException("A clip must have a name.");
        }

        if (sourceDuration is { } duration)
        {
            segment.Range.ValidateAgainst(duration);
        }

        if (_segments.Any(existing => existing.Id != replacing && existing.Overlaps(segment)))
        {
            throw new InvalidDataException("The clip overlaps an existing retained source range.");
        }
    }

    public bool Equals(EditList? other) =>
        other is not null && _segments.SequenceEqual(other._segments);

    public override bool Equals(object? obj) => obj is EditList other && Equals(other);

    public override int GetHashCode()
    {
        var hash = new HashCode();
        foreach (var segment in _segments)
        {
            hash.Add(segment);
        }

        return hash.ToHashCode();
    }
}
